src/seq_mv/csr_matvec.c

```diff
@@ -209,11 +212,303 @@ hypre_CSRMatrixMatvecOutOfPlaceHost( HYPRE_Complex    alpha,
+            }
+         } // y = -A*x - b
+         else
+         {
+#ifdef HYPRE_USING_OPENMP
+#pragma omp parallel for private(i) HYPRE_SMP_SCHEDULE
+#endif
+            for (i = 0; i < num_rows; i++)
+            {
+               y_data[i] = alpha*b_data[i];
+            }
+
+#ifdef HYPRE_USING_OPENMP
+#pragma omp parallel for private(i,j,m,tempx) HYPRE_SMP_SCHEDULE
+#endif
+            for (i = 0; i < num_rownnz; i++)
+            {
+               m = A_rownnz[i];
+               tempx = 0.0;
+               for (j = A_i[m]; j < A_i[m+1]; j++)
+               {
+                  tempx += A_data[j] * x_data[A_j[j]];
+               }
+               y_data[m] = alpha*(tempx + b_data[m]);
+            }
+         } // y = alpha*(A*x + b)
+      }
+      else
+      {
+         if (alpha == 1.0)
+         {
+#ifdef HYPRE_USING_OPENMP
+#pragma omp parallel for private(i) HYPRE_SMP_SCHEDULE
+#endif
+            for (i = 0; i < num_rows; i++)
+            {
+               y_data[i] = beta*b_data[i];
+            }
+
+#ifdef HYPRE_USING_OPENMP
+#pragma omp parallel for private(i,j,m,tempx) HYPRE_SMP_SCHEDULE
+#endif
+            for (i = 0; i < num_rownnz; i++)
+            {
+               m = A_rownnz[i];
+               tempx = 0.0;
+               for (j = A_i[m]; j < A_i[m+1]; j++)
+               {
+                  tempx += A_data[j] * x_data[A_j[j]];
+               }
+               y_data[m] = tempx + beta*b_data[m];
+            }
+         } // y = A*x + beta*b
+         else if (-1 == alpha)
+         {
+#ifdef HYPRE_USING_OPENMP
+#pragma omp parallel for private(i) HYPRE_SMP_SCHEDULE
+#endif
+            for (i = 0; i < num_rows; i++)
+            {
+               y_data[i] = -temp*b_data[i];
+            }
+
+#ifdef HYPRE_USING_OPENMP
+#pragma omp parallel for private(i,j,m,tempx) HYPRE_SMP_SCHEDULE
+#endif
+            for (i = 0; i < num_rownnz; i++)
+            {
+               m = A_rownnz[i];
+               tempx = 0.0;
+               for (j = A_i[m]; j < A_i[m+1]; j++)
+               {
+                  tempx -= A_data[j] * x_data[A_j[j]];
+               }
+               y_data[m] = tempx - temp*b_data[m];
+            }
+         } // y = -A*x - temp*b
+         else
+         {
+#ifdef HYPRE_USING_OPENMP
+#pragma omp parallel for private(i) HYPRE_SMP_SCHEDULE
+#endif
+            for (i = 0; i < num_rows; i++)
+            {
+               y_data[i] = beta*b_data[i];
+            }
+
+#ifdef HYPRE_USING_OPENMP
+#pragma omp parallel for private(i,j,m,tempx) HYPRE_SMP_SCHEDULE
+#endif
+            for (i = 0; i < num_rownnz; i++)
+            {
+               m = A_rownnz[i];
+               tempx = 0.0;
+               for (j = A_i[m]; j < A_i[m+1]; j++)
+               {
+                  tempx += A_data[j] * x_data[A_j[j]];
+               }
+               y_data[m] = alpha*tempx + beta*b_data[m];
+            }
+         } // y = alpha*(A*x + temp*b)
+      } // temp != 0 && temp != -1 && temp != 1
+   }
```

Comment:
Aren't you adding beta*b_data[i] here twice for the nonzero A rows?

Reply:
Thanks for the suggestion! `y_data[m] += alpha*tempx` is actually better. I will make the change